### DISCUSSION TITLE
HTTP vs HTTPS for URIs

### DISCUSSION THREAD

**Discussion by bilalshaikh42** [biosimulations/biosimulations#3973] [Discussions]
I was looking into the use of HTTP vs HTTPS for URIs. As everything generally moves to HTTPS I had initially thought it would make sense for us to prefer HTTPS URIs over HTTP, especially when those URIs are embedded into a web page. Most web browsers and code quality/web ranking tools discourage the use of HTTP links.  

However, there is quite a lot of legacy information/resources that are reliant on the HTTP prefix.  Ideally, any links that are embedded as HTTPS would have automatic redirecting built into their infrastructure.  For example, http://biosimulations.org will automatically redirect to the HTTPS version.  Unfortunately, this is not the case for some of the resources that we use. For example, no HTTPS version of http://biomodels.net/biology-qualifiers/encodes  exists. 

In any case,  here are some relevant resources that I found while looking into this

- https://www.w3.org/blog/2016/05/https-and-the-semantic-weblinked-data/  
  Argues for using HTTP and letting infrastructure redirect automatically 

- https://hal.inria.fr/hal-01673291/document
 Argues that this is dangerous and just HTTPS should be used

Some issues that have popped up when trying to change things over to HTTPS: 

- https://github.com/schemaorg/schemaorg/issues/2853
- https://github.com/schemaorg/schemaorg/issues/2852
- https://forum.geonames.org/gforum/posts/list/37583.page

I agree with our current recommendation of using HTTP universally for URIs, though I would suggest changing the URIs to HTTPS specifically when embedding `identifiers.org `URLs into web pages since we are sure that an HTTPS version exists. This would not be a recommendation for users to change the URIs in the metadata, but a simple angular pipe / if statement that changes the link embedded on a particular page. Not a high priority issue, but might help avoid some security warnings over the longer term

**Comment by jonrkarr** [comment]
While structurally similar, URIs and URLs a few key differences:
- URIs are mainly intended to be unique strings.
- URIs aren't necessarily intended to be resolvable like URLs.
- Since URIs aren't necessarily intended to be resolved, `https` is mostly relevant to URLs.

Regarding identifiers.org, it forwards to the registered URL pattern for each namespace, which is `http` or `https`, as appropriate for each namespace. For example, http://identifiers.org/pubmed:1234 forwards to https://pubmed.ncbi.nlm.nih.gov/1234/. Basically, identifiers.org already takes care of mapping `http` to `https` as relevant to each namespace. I don't believe identifiers.org requires prefixes to be mapped to `https`. I believe they support `http` and `https`.

Since identifiers.org already forwards to `https` as appropriate, the benefit of us changing identifiers links from `http` to `https` seems minimal. @bilalshaikh42 do you feel this is important?

**Reply by bilalshaikh42** [reply]
Not super important no. Just something that would be good to do when creating a page/component. No need to go through and change all the existing ones.  When the redirect takes some time due to a slow connection etc, browsers display a "not secure" warning. Based on recent trends, this will probably be made more prominent over time. Its just a small UI thing

**Reply by jonrkarr** [reply]
The relevant redirect time is just the time it takes identifiers.org to resolve an id. I think this is a pretty stable resource.

I think we can convert all metadata hyperlinks involving http://identifiers.org to https at 1 line of code.

**Reply by jonrkarr** [reply]
I'll commit a change which converts all metadata hyperlinks involving http://identifiers.org to https.

**Reply by jonrkarr** [reply]
Another common resource we could automatically convert from http to https is purl.org. If there are others, we can support them too. However, I would prefer to register namespaces with identifiers.org rather than encourage purl.org. They fill similar roles but for different domains.
